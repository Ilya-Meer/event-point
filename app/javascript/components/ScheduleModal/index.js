import React, { Fragment, useState, forwardRef, useRef } from 'react';
import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';
import Modal from 'react-bootstrap/Modal';
import Dropdown from 'react-bootstrap/Dropdown';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';

const ScheduleModal = ({
  show,
  handleDismiss,
  events,
  handleScheduleEvent,
}) => {
  const [selectedEventState, setSelectedEventState] = useState({
    datetime: new Date(),
  });

  const handleChangeScheduledEvent = (key, value) => {
    setSelectedEventState({ ...selectedEventState, [key]: value });
  };

  const ref = useRef(null);

  const CustomDatePickerInput = forwardRef(({ onClick, value }, ref) => (
    <Fragment>
      <label htmlFor='schedule-datepicker-input'>Date &amp; Time</label>
      <span className='form-input-required'>*</span>
      <input
        id='schedule-datepicker-input'
        className='schedule-date-input'
        onClick={onClick}
        value={value}
        onChange={onClick}
        ref={ref}
      />
    </Fragment>
  ));

  const renderDropdownDisplay = () => {
    const defaultMessage = 'Select Event';

    if (!events || !events.length) {
      return defaultMessage;
    }

    const selectedEvent = events.find(
      (event) => event.id === Number(selectedEventState.event_id)
    );

    if (!selectedEvent) {
      return defaultMessage;
    }

    return selectedEvent.topic;
  };

  return (
    <Modal show={show} onHide={handleDismiss} className='modal'>
      <Modal.Header closeButton>
        <Modal.Title>Schedule Event</Modal.Title>
      </Modal.Header>
      <Modal.Body>Select from existing events</Modal.Body>
      <Form
        className='modal-form'
        onSubmit={(e) => {
          e.preventDefault();

          handleScheduleEvent(selectedEventState);
          setSelectedEventState({ datetime: new Date() });
        }}
      >
        <Dropdown>
          <Dropdown.Toggle
            className='schedule-modal-dropdown'
            data-testid='schedule-dropdown-toggle'
          >
            {renderDropdownDisplay()}
          </Dropdown.Toggle>
          <Dropdown.Menu className='schedule-dropdown-menu'>
            {events.map((event) => (
              <Dropdown.Item
                key={event.id}
                onSelect={(event_id) =>
                  handleChangeScheduledEvent('event_id', event_id)
                }
                eventKey={event.id}
              >
                {event.topic}
              </Dropdown.Item>
            ))}
          </Dropdown.Menu>
        </Dropdown>
        <DatePicker
          selected={selectedEventState.datetime}
          onChange={(date) => handleChangeScheduledEvent('datetime', date)}
          showTimeSelect
          minDate={new Date()}
          dateFormat='MMM d, yyyy h:mm aa'
          timeFormat='h:mm aa'
          customInput={<CustomDatePickerInput ref={ref} />}
        />
        <Modal.Footer className='modal-footer'>
          <Button
            variant='primary'
            type='submit'
            className='modal-submit'
            data-testid='schedule-event-button'
          >
            Schedule Event
          </Button>
          <Button onClick={handleDismiss} variant='secondary' type='button'>
            Close
          </Button>
        </Modal.Footer>
      </Form>
    </Modal>
  );
};

export default ScheduleModal;

ScheduleModal.propTypes = {
  show: PropTypes.bool,
  handleDismiss: PropTypes.func,
  events: PropTypes.array,
  handleScheduleEvent: PropTypes.func,
};
