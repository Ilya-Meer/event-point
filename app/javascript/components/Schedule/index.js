import React, { Fragment, useState, useMemo } from 'react';
import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import ScheduleModal from '../ScheduleModal';
import { scheduleEvent, getScheduledEvents } from '../../utils/api';
import { convertToSchedule } from '../../utils/date';

const renderDay = (scheduledDay, eventsForDay) => {
  return (
    <Fragment key={scheduledDay}>
      <div className='scheduled-day'>{scheduledDay}</div>
      <ul>
        {eventsForDay.map((event) => (
          <li key={event.created_at} className='scheduled-event'>
            {event.topic}
          </li>
        ))}
      </ul>
    </Fragment>
  );
};

const Schedule = ({
  allEvents,
  scheduledEvents,
  updateSchedule,
  updateAllEvents,
}) => {
  const [showScheduleModal, setShowScheduleModal] = useState(false);

  const [formattedSchedule, setFormattedSchedule] = useState([]);

  useMemo(() => {
    let formatted = [];
    for (let [date, events] of Object.entries(scheduledEvents)) {
      formatted.push(renderDay(date, events));
    }

    setFormattedSchedule(formatted);
  }, [scheduledEvents]);

  const handleScheduleEvent = async (eventState) => {
    try {
      await scheduleEvent({
        ...eventState,
        datetime: new Date(eventState.datetime.toUTCString()),
      });

      const updated = await getScheduledEvents().then((res) => res.json());
      updateSchedule(convertToSchedule(updated));

      const newEventsArr = allEvents.map((event) => {
        if (event.id === Number(eventState.event_id)) {
          return { ...event, datetime: eventState.datetime };
        }

        return event;
      });

      updateAllEvents(newEventsArr);

      setShowScheduleModal(false);
    } catch (error) {
      console.error(error);
    }
  };

  return (
    <Fragment>
      <div className='schedule-heading'>
        <h3>Scheduled Events</h3>
        <Button onClick={() => setShowScheduleModal(true)}>
          + Schedule Event
        </Button>
      </div>
      <div className='schedule-wrapper'>{formattedSchedule}</div>
      <ScheduleModal
        show={showScheduleModal}
        handleDismiss={() => setShowScheduleModal(false)}
        events={allEvents}
        handleScheduleEvent={handleScheduleEvent}
      />
    </Fragment>
  );
};

Schedule.propTypes = {
  allEvents: PropTypes.array,
  scheduledEvents: PropTypes.object,
  scheduleEvent: PropTypes.func,
};

Schedule.defaultProps = {
  scheduledEvents: {},
};

export default Schedule;
