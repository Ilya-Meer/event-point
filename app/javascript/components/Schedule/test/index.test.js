import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent, act } from '@testing-library/react';
import Schedule from '..';
import ScheduleModal from '../../ScheduleModal';

it('renders without crashing', () => {
  const tree = renderer
    .create(
      <Schedule
        allEvents={[]}
        scheduledEvents={{}}
        updateSchedule={() => {}}
        updateAllEvents={() => {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('brings up the modal', async () => {
  render(
    <Schedule
      allEvents={[]}
      scheduledEvents={{}}
      updateSchedule={() => {}}
      updateAllEvents={() => {}}
    />
  );

  await act(async () => {
    await fireEvent.click(screen.getByText(/Schedule Event/));
  });

  expect(screen.queryByText('Select from existing events')).toBeInTheDocument();
});

it('schedules an event', async () => {
  const allEvents = [
    {
      id: 5,
      topic: 'An interesting topic',
    },
    {
      id: 6,
      topic: 'Another interesting topic',
    },
  ];

  render(
    <Schedule
      allEvents={allEvents}
      scheduledEvents={{}}
      updateSchedule={() => {}}
      updateAllEvents={() => {}}
    />
  );

  await act(async () => {
    await fireEvent.click(screen.getByText(/Schedule Event/));
  });

  await act(async () => {
    await fireEvent.click(screen.getByTestId('schedule-dropdown-toggle'));
  });

  const firstEvent = screen.queryByText('An interesting topic');

  expect(firstEvent).toBeInTheDocument();

  await act(async () => {
    await fireEvent.click(firstEvent);
  });

  const datePickerInput = screen.queryByRole('textbox');
  expect(datePickerInput).toBeInTheDocument();

  // TODO: Research mocking current date in order to correctly simulate datepicker selection

  fetchMock.once(JSON.stringify({})).once(JSON.stringify([]));

  await act(async () => {
    await fireEvent.click(screen.getByTestId('schedule-event-button'));
  });

  const requestBody = JSON.parse(fetchMock.mock.calls[0][1].body);
  expect(requestBody.event_id).toEqual(allEvents[0].id.toString());

  expect(fetchMock.mock.calls[0][0]).toEqual(
    'http://localhost:3000/schedule_event'
  );
});
