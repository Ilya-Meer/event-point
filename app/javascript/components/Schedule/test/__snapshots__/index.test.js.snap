// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Array [
  <div
    className="schedule-heading"
  >
    <h3>
      Scheduled Events
    </h3>
    <button
      className="btn btn-primary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      + Schedule Event
    </button>
  </div>,
  <div
    className="schedule-wrapper"
  />,
]
`;
